automation/taskcluster/lib/taskgraph.py

```diff
@@ -0,0 +1,22 @@
+import taskcluster
+
+from lib import tasks
+
+
+class TaskGraph:
```

Comment:
I believe a class is unnecessary, for now: this can be solved in a single function https://github.com/mozilla-mobile/fenix/blob/f9c791c93b64c71328f425e8af192e95f37e70db/automation/taskcluster/lib/tasks.py#L504.

Reply:
Without this class, you need to manage a bunch of dictionaries in your `nightly()`/`master_push()`/`on_pr_or_push()` functions, which is annoying. This class captures all that logic in one place, allowing the scheduling functions to be as simple as possible :)

Reply:
I'm sorry, I don't think I agree.

This class buries IO https://github.com/mozilla-mobile/reference-browser/pull/681/files/83032d77133915e9e1432b75aff025f8080d2d15#diff-5c81e10cf0ba5116a7f050bb076f174fR13, which is not nice from a "clean architecture" standpoint: https://www.youtube.com/watch?v=DJtef410XaM

Moreover, the way this function is called https://github.com/mozilla-mobile/reference-browser/pull/681/files/83032d77133915e9e1432b75aff025f8080d2d15#diff-168d551d2c0830884a5caeb50ae576e9R337 makes the data flow like this:

```
process task payload => IO => process another task payload => IO => [...]
``` 

Resolving the scheduling this way is not ideal, because any bug in the process part would result in a partially submitted graph. If we reduce the flow to 

```
process all tasks payload => IO => Done
``` 

then, the risk of having a partially submitted graph is lower. 


The scheduling can still happen at a single place, and the scheduling function can be really simple too. What do you think?

Reply:
Hmm, interesting point about IO :thinking: I wonder to correctly apply the idea of "avoiding burying IO" - if we avoid putting any abstractions around our IO, what about simple things like `urllib` or, more complex-ly, Taskcluster's `queue` abstraction? Is `queue` burying IO?
My personal interpretation is that IO shouldn't be mixed with domain logic. I think that abstractions that are specifically for IO (such as for the batching of task-scheduling) aren't bad, eh? Hmm :thinking: 

-----

Ah, good point about partially-submitted graphs! Batching those make a **ton** of sense!

Perhaps there's still value in easing the work of the consumer by handling the batching with an abstraction? I think it's too verbose to have to keep track of all the ids and order from `nightly(...)`, and it would be ideal if I could just call `.add_task(...)` multiple times, then calling `.schedule(...)` at the end would send them all to Taskcluster and return a raw graph?


:thinking: Interested in your thoughts!

Reply:
I agree, keeping track of IDs is cumbersome and eventually unnecessary. Taskgraph solves this by providing labels and then, at the end of the pipeline, it translates them to actual taskIds. In any case, we need a way to link one task to another (by ids or labels or something else). I'm looking forward to the day we have taskgraph on mobile :+1: 
In the mean time, I think it's okay to keep ids in `nightly()` and `pr_or_push()`. How about we reevaluate this whenever `nightly()`/`pr_or_push()` become too complex (and taskgraph is not ready yet)?

Regarding, burying IO, I agree with you, it's better to have some abstractions decoupled from the data we want to IO. I think using some layer of abstractions (like Taskcluster's `queue`) is great: that's an external library and we can test our code by just mocking it out. What worried me about this class was the timing of the IO, which is not really the point of clean architecture. I looked back at https://github.com/mozilla-mobile/fenix/blob/f9c791c93b64c71328f425e8af192e95f37e70db/automation/taskcluster/lib/tasks.py#L504 and this function does the same in terms of abstractions. Sorry about this, I used a bad argument.

Reply:
No worries, yeah, sounds good!

I think I might've been ambiguous, too: I completely understand the need for us to handle ids - regardless of abstractions, you still need to manage references between data. To be clear, I'm not worried at all about keeping track of `build_task_id` + friends :)

However, I _do_ think that it's cumbersome to [keep track of all the sets of tasks](https://github.com/mozilla-mobile/reference-browser/pull/681/files#diff-168d551d2c0830884a5caeb50ae576e9R95) and [return this specially-grouped data](https://github.com/mozilla-mobile/reference-browser/pull/681/files#diff-168d551d2c0830884a5caeb50ae576e9R91) (we're leaking information about the build+sign+other lifecycle to the caller, who shouldn't need to care - what happens when a new `pr_or_push_but_different()` function is needed that doesn't include signing tasks? That function would need to return a dummy `{}`, which would be confusing and noisy)

That logic exists to handle batching of tasks, and I think that it can safely be abstracted away to remove "noise" (imho) from the `nightly()`/`pr_or_push()`/etc functions.

For example, instead of:
```
assemble_tasks = {}
sign_tasks = {}

assemble_task_id = generate_id()
assemble_tasks[assemble_task_id] = craft_assemble_task()
# ... multiple times

sign_tasks[generate_id()] = craft_sign_task(assemble_task_id)
# ... multiple times

return (assemble_tasks, sign_tasks, ...)
```
I'm concerned about the `$group_tasks` dictionaries, and I think we can get the best-of-both-worlds if we:
```
# task_scheduler = ...
assemble_task_id = task_scheduler.add_task(craft_assemble_task())
# ... multiple times
sign_task_id = task_scheduler.add_task(craft_sign_task(assemble_task_id))
# ... multiple times
return task_scheduler.apply() # schedules tasks, returns raw_graph
```

Reply:
Yeah, that can work. I agree the dummy `{}` is yuck. Nonetheless, the class model makes it harder to see what's the data flow (per the "clean architecture" described in the video above). You have an object that has a state, instead of having a full set of tasks that you pass to a schedule function. I think the best solution would be to have data class, that doesn't know anything about scheduling, and you pass that instance (of the data class) to a schedule function. This way, you can stop at a step of the data flow, and now what you're about to schedule.

Let's do it in a follow up.